### PR TITLE
Refactor controllers to avoid boilerplate code

### PR DESCRIPTION
**Pull request by schweikart**
This pull request replaces our controllers with a generic router implementation that discovers and routes new meta solvers (problem types) automatically.

### CODE REVIEW

**Comment by schweikart** [comment]
Feature Model tests are currently failing because the feature model anomaly solver currently does not use the same route structure as the other solvers. We can solve this problem by separating all feature model anomalies into different problem types (implemented in #40 ) or by combining the feature model anomaly routes into a single route and adding an `AnomalyType` parameter to  `FeatureModelAnomalyRequest`.

**Comment by Elscrux** [comment]
Cross Origin is still a problem here, and we'll have to sort that out for the new webflux framework.

File: src/main/java/edu/kit/provideq/toolbox/api/SolveRouter.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                                                                ServerRequest req) {
+    var solution = req.queryParam("id")
+        .map(Long::parseLong)
+        .map(solutionId -> metaSolver.getSolutionManager().getSolution(solutionId))
+        .map(Solution::toStringSolution)
+        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
+            "Could not find a solution for this problem with this solution id!"));
+
+    // yes, solution is of type `Solution<String>`.
+    // No idea why `toStringSolution` returns `SolutionHandle`
```
Comment (Elscrux) [inline src/main/java/edu/kit/provideq/toolbox/api/SolveRouter.java:139]: Probably in a misguided effort to use the type with the fewest assumptions. Feel free to change that.

**Review by Elscrux — CHANGES_REQUESTED** [review]
Looks good so far! The functional approach really shines here.
I think we could move the documentation part to its own method in each class, similarly to the handle method. That would separate this part a bit, reduce nesting and the method name would also help to document what this is for (I first didn't really get what this is about until you told me).

**Comment by schweikart** [comment]
> Looks good so far! The functional approach really shines here. I think we could move the documentation part to its own method in each class, similarly to the handle method. That would separate this part a bit, reduce nesting and the method name would also help to document what this is for (I first didn't really get what this is about until you told me).

That seems like a good idea! Can you address this in your PR for API documentation?

**Comment by Elscrux** [comment]
> Cross Origin is still a problem here, and we'll have to sort that out for the new webflux framework.

Tested again with the latest versions of the web (merged PR) and toolbox (anomaly refactor), this should have been fixed.

**Comment by Elscrux** [comment]
> > Looks good so far! The functional approach really shines here. I think we could move the documentation part to its own method in each class, similarly to the handle method. That would separate this part a bit, reduce nesting and the method name would also help to document what this is for (I first didn't really get what this is about until you told me).
> 
> That seems like a good idea! Can you address this in your PR for API documentation?

Yeah I can do that as part of the API documentation.

File: src/main/java/edu/kit/provideq/toolbox/api/SolveRouter.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    return metaSolverProvider.getMetaSolvers().stream()
+        .map(this::defineRouteForMetaSolver)
+        .reduce(RouterFunction::and)
+        .orElseThrow(); // we should always have at least one route or the toolbox is useless
+  }
+
+  private RouterFunction<ServerResponse> defineRouteForMetaSolver(MetaSolver<?, ?, ?> metaSolver) {
+    String problemId = metaSolver.getProblemType().getId();
+    return route().POST(
+        "/solve/" + problemId,
```
[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/api/SolveRouter.java]: This is still a thing right? You just have two function calls right now.

[the remark this answers is not in the document]
Comment (Elscrux) [reply src/main/java/edu/kit/provideq/toolbox/api/SolveRouter.java]: Nevermind, in my API refactoring I'll need to call things from two different methods anyway. And this is not a huge deal. It's good that the magic string is only used once in a central place now.

**Review by Elscrux — APPROVED** [review]
This all looks fine to me now, but I'll wait for your SolutionHandle removal (if you want to do that, and do it in this PR)

**Comment by schweikart** [comment]
> This all looks fine to me now, but I'll wait for your SolutionHandle removal (if you want to do that, and do it in this PR)

I'll do that in another PR but please merge #40 before merging this one :)

**Pull request merged**